indica/readers/abstractreader.py

```diff
@@ -84,985 +44,187 @@ def __init__(
+        if instrument not in self.instrument_methods.keys():
             raise ValueError(
                 "{} does not support reading for instrument {}".format(
                     self.__class__.__name__, instrument
                 )
             )
-        method = getattr(self, self.INSTRUMENT_METHODS[instrument])
-        if not quantities:
-            quantities = set(self.available_quantities(instrument))
-        return method(uid, instrument, revision, quantities, **kwargs)
-
-    def get_thomson_scattering(
-        self,
-        uid: str,
-        instrument: str,
-        revision: RevisionLike,
-        quantities: Set[str],
-        **kwargs,
-    ) -> Dict[str, DataArray]:
-        """
-        Reads data based on Thomson Scattering.
-        """
-        database_results = self._get_thomson_scattering(
-            uid, instrument, revision, quantities
-        )
-
-        transform = TransectCoordinates(
-            database_results["x"],
-            database_results["y"],
-            database_results["z"],
-            f"{instrument}",
-            machine_dimensions=database_results["machine_dims"],
-        )
-        database_results["channel"] = np.arange(database_results["length"])
+        # Read data from database
+        data = self._read_database(uid, instrument, revision)
+
+        # Re-arrange data (machine-specific)
+        method = self.instrument_methods[instrument]
+        data = getattr(self, f"_{method}")(data)
+
+        if self.return_dataarrays:
+            # Instatiate transforms
+            if "location" in data and "direction" in data:
+                transform = LineOfSightTransform(
+                    data["location"][:, 0],
+                    data["location"][:, 1],
+                    data["location"][:, 2],
+                    data["direction"][:, 0],
+                    data["direction"][:, 1],
+                    data["direction"][:, 2],
+                    machine_dimensions=data["machine_dims"],
+                    dl=dl,
+                    passes=passes,
+                )
+            elif "x" in data and "y" in data and "z" in data:
+                transform = TransectCoordinates(
+                    data["x"],
+                    data["y"],
+                    data["z"],
+                    machine_dimensions=data["machine_dims"],
+                )
+            else:
+                transform: CoordinateTransform = TrivialTransform
 
-        data = {}
-        dims = ["t", "channel"]
-        for quantity in quantities:
-            if database_results.get(quantity) is None:
-                continue
-            data[quantity] = self.assign_dataarray(
-                instrument,
-                quantity,
-                database_results,
-                dims,
-                transform=transform,
+            # Build data-arrays
+            quantities = READER_QUANTITIES[method]
+            data = build_dataarrays(
```

Comment:
And here too. There should be an intermediate step that just simply gets the data and does nothing to it.

Reply:
I don't agree with this.
Returning the data to me means either of the following:
- return the database results dictionary
- return the data arrays with associated transform

If transform is None then it's simply not assigned as attribute, which is the case for the data_arrays that are created in the plasma class